### DISCUSSION TITLE
Tags Subscription Doubts

### DISCUSSION THREAD

**Discussion by yasodakrishnav** [eclipse-milo/milo#1146] [General]
Hey Team,

1, How can I efficiently subscribe to and unsubscribe the tags in bulk?
2. Is it possible to disconnect the client and still receive data change notifications after subscribing to tags? if not how to can i maintain that client?
3. If I need to subscribe to additional tags on the same server, how can I accomplish that?
4. Could you provide clarification on the concepts of sampling and publishing intervals?
5. What steps should I take to maintain the subscription list on the server in the event of a server restart?
6. What is the meaning of MonitoringMode.Sampling and MonitoringMode.Reporting?
7. Which attribute indicates whether a tag is writable or not?

Would be help full if you can provide the code examples. 

Thanks
@kevinherron

**Comment by kevinherron** [comment]
> How can I efficiently subscribe to and unsubscribe the tags in bulk?

Use the plural create monitored items call on either `ManagedSubscription` or `UaSubscription`. 
https://github.com/eclipse/milo/blob/c685a5b0251ee38c0098fa7050417d1593c7d249/opc-ua-sdk/sdk-client/src/main/java/org/eclipse/milo/opcua/sdk/client/subscriptions/ManagedSubscription.java#L276-L291
https://github.com/eclipse/milo/blob/c685a5b0251ee38c0098fa7050417d1593c7d249/opc-ua-sdk/sdk-client/src/main/java/org/eclipse/milo/opcua/sdk/client/api/subscriptions/UaSubscription.java#L111-L130

>Is it possible to disconnect the client and still receive data change notifications after subscribing to tags? if not how to can i maintain that client?

No... not sure how you would expect that to work. The client will automatically reconnect if the connection is lost, however.

>If I need to subscribe to additional tags on the same server, how can I accomplish that?

Just create more monitored items in the same way you created the first ones.

>Could you provide clarification on the concepts of sampling and publishing intervals?

Sampling Interval is how often the underlying data for each Monitored Item is sampled. Publishing Interval is how often a Subscription "executes", gathering up any queued data changes for Monitored Items and sending them all back to the client via a Publish Response.
https://reference.opcfoundation.org/Core/Part4/v105/docs/5.12.1
https://reference.opcfoundation.org/Core/Part4/v105/docs/5.13.1

>What steps should I take to maintain the subscription list on the server in the event of a server restart? 

Add a `SubscriptionListener` to `UaSubscriptionManager`. You will get a callback to `SubscriptionListener::onSubscriptionTransferFailed` if a Subscription can't be transferred from old Session to new Session when client reconnects. In this case you must create the Subscription and Monitored Items again. This will require your own bookkeeping to know what tags you need to resubscribe to.

> What is the meaning of MonitoringMode.Sampling and MonitoringMode.Reporting?

`Sampling` mode means the server is still sampling the item and queuing changes but not reporting them to the client. `Reporting` means they are sampling _and_ reporting them when the Subscription publishes.

> Which attribute indicates whether a tag is writable or not?

`AttributeId.AccessLevel` and `AttributeId.UserAccessLevel`.

**Reply by yasodakrishnav** [reply]
Thanks @kevinherron  for the above response.
             
              `OpcUaClient client = getOrCreateClient(uaController);
                ManagedSubscription managedSubscription = ManagedSubscription.create(client);
                List<NodeId> nodeIds =  getNodeIdList();

                List<ManagedDataItem> dataItems = managedSubscription.createDataItems(nodeIds);
                for (ManagedDataItem dataItem : dataItems) {
                    if(dataItem.getStatusCode().isGood()){
                        String identifier = dataItem.getNodeId().getIdentifier().toString();
                    }
                }`
                
                1. How can i receive the change of value notification to ManagedSubscription? 
                2. Could you provide the code to fetch Engineering Unit for analog type of nodes, i need to do while browsing the nodes.?

**Reply by yasodakrishnav** [reply]
Hey @kevinherron  

Could you help he with the above doubt?

Thanks

**Comment by kevinherron** [comment]
> 1. How can i receive the change of value notification to ManagedSubscription? 

This is covered in the examples: https://github.com/eclipse/milo/blob/c685a5b0251ee38c0098fa7050417d1593c7d249/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/ManagedSubscriptionDataExample.java#L39-L46

https://github.com/eclipse/milo/wiki/Client#managedsubscription

> 2. Could you provide the code to fetch Engineering Unit for analog type of nodes, i need to do while browsing the nodes.?

Probably need to know more about your browsing code. Engineering Units is a separate Property Node that would show up in the browse underneath an analog type Node, so you would need to keep track of what you're browsing so that you understand the child Property Nodes belong to some parent.

**Reply by yasodakrishnav** [reply]
Alright kevin, Thanks

**Comment by yasodakrishnav** [comment]
Hi kevin,

I have node Ids, now How can i remove the subscribed tags using NodeId?

**Reply by kevinherron** [reply]
You can't delete a MonitoredItem having just a NodeId. When you create a MonitoredItem it gets assigned an id by the server. This id must be referenced during the call to delete MonitoredItems as well.

**Reply by yasodakrishnav** [reply]
Is there anyway to get the MonitoredItems from opc ua client using node ids to delete?

**Reply by kevinherron** [reply]
All of the MonitoredItems you've created can be retrieved by accessing the SubscriptionManager via `OpcUaClient::getSubscriptionManager`, then getting the subscriptions, then getting the items from each subscription. This is assuming you're not doing your own bookkeeping already with the items that get created.

It's possible to create multiple MonitoredItems, for multiple attributes, for each NodeId. It's your applications responsibility to figure out which items to delete.

**Reply by yasodakrishnav** [reply]
Let me explain my scenario kevin.

Subscription :

I will subscribe the points like using below method

`private List<ManagedDataItem> subscribePoints(OpcUaController uaController, List<OpcPoint> points) throws UaException {
        OpcUaClient client = OpcUtils.getCachedOrCreateOpcUaClient(uaController);
        ManagedSubscription managedSubscription = ManagedSubscription.create(client);
        managedSubscription.addDataChangeListener(this::processSubscriptionData);
        List<NodeId> nodeIds = OpcUtils.getNodeIdList(points);
        return managedSubscription.createDataItems(nodeIds);
    }`
    
    * OpcUaController Object contains fields required to create the OpcUaClient.
    * I will subscribe for Value attribute only, so Can i use the same managed subscription to subscribe points from next time?
   * I will call this method  again later if i want to subscribe some more points using cached opcUaClient.
   * Now i will only have point namespace Index and point identifier to unsubscribe the points.
   
   
   would be great if you could optimize the code/ provide some suggestions.
   
   Thanks

**Reply by kevinherron** [reply]
This method returns a `List<ManagedDataItem>` - what are you doing with these? These are everything you need to delete them later. They even have a delete method on them.

**Reply by yasodakrishnav** [reply]
could you mention the method name?

**Reply by kevinherron** [reply]
There are `delete` and `deleteAsync` methods. It is also shown in the example: https://github.com/eclipse/milo/blob/master/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/ManagedSubscriptionDataExample.java#L58

**Reply by yasodakrishnav** [reply]
alright kevin, thanks for your time.
